### PR TITLE
Migrate Read in Core to JUnit5

### PR DESCRIPTION
**Pull request by tomtongue**
Migrate the following "Read" in iceberg-core to JUnit 5 and AssertJ style for https://github.com/apache/iceberg/issues/9085.

## Current Progress (same as in the issue above)
Read:
- [x] `MetadataTableScanTestBase` (extends `TableTestBase`) 
  - [x] `TestMetadataTableScans` 
  - [x] `TestMetadataTableScansWithPartitionEvolution`
- [x] `TestEntriesMetadataTable`
- [x] `TestBatchScans`
- [x] `TestIncrementalDataTableScan`
- [x] `TestFindFiles`

### CODE REVIEW

**Comment by tomtongue** [comment]
@nastra Complete the migration of the above classes. Could you review this when you have time?

File: core/src/test/java/org/apache/iceberg/MetadataTableScanTestBase.java [hunk 1]
```diff
@@ -89,18 +88,16 @@ protected void validatePartition(
...
-            .anyMatch(
-                entry -> {
-                  StructLike partition = entry.file().partition();
-                  if (position >= partition.size()) {
-                    return false;
-                  }
-
-                  return Objects.equals(partitionValue, partition.get(position, Object.class));
-                }));
+    assertThat(StreamSupport.stream(entries.spliterator(), false))
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/MetadataTableScanTestBase.java:91]:
```suggestion
    assertThat(entries)
```
I think this should also work

File: core/src/test/java/org/apache/iceberg/MetadataTableScanTestBase.java [hunk 2]
```diff
@@ -89,18 +88,16 @@ protected void validatePartition(
...
+        .as("File scan tasks do not include correct file")
+        .anyMatch(
+            entry -> {
+              StructLike partition = entry.file().partition();
+              if (position >= partition.size()) {
+                return false;
+              }
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/MetadataTableScanTestBase.java:98]: nit: please add newline after }

File: core/src/test/java/org/apache/iceberg/TestBatchScans.java [hunk 3]
```diff
@@ -51,27 +48,27 @@ public void testDataTableScan() {
...
-    Assert.assertEquals("Expected 2 tasks", 2, tasks.size());
+    assertThat(tasks).hasSize(2);
 
     FileScanTask t1 = tasks.get(0).asFileScanTask();
-    Assert.assertEquals("Task file must match", t1.file().path(), FILE_A.path());
-    V1Assert.assertEquals("Task deletes size must match", 0, t1.deletes().size());
-    V2Assert.assertEquals("Task deletes size must match", 1, t1.deletes().size());
+    assertThat(FILE_A.path()).as("Task file must match").isEqualTo(t1.file().path());
+    V1Assert.assertHasSize("Task deletes size must match", 0, t1.deletes());
+    V2Assert.assertHasSize("Task deletes size must match", 1, t1.deletes());
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestBatchScans.java:56]: I think changing this should be done separately from migrating away from JUnit4. That means that we can consider making `TableAssertions` more fluent and more like AssertJ assertions, but that shouldn't be part of this PR

File: core/src/test/java/org/apache/iceberg/TestBase.java [hunk 4]
```diff
@@ -745,6 +745,12 @@ void assertEquals(String context, Object expected, Object actual) {
         assertThat(actual).as(context).isEqualTo(expected);
       }
     }
+
+    void assertHasSize(String context, int expected, List<?> actualList) {
```
[the remark this answers is not in the document]
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestBase.java:749]: as I mentioned in another comment, I don't think we should introduce this method as part of this PR

File: core/src/test/java/org/apache/iceberg/TestEntriesMetadataTable.java [hunk 5]
```diff
@@ -68,20 +63,19 @@ public void testEntriesTableScan() {
...
     FileScanTask file = Iterables.getOnlyElement(scan.planFiles());
-    Assert.assertEquals(
-        "Data file should be the table's manifest",
-        Iterables.getOnlyElement(table.currentSnapshot().allManifests(table.io())).path(),
-        file.file().path());
-    Assert.assertEquals("Should contain 2 data file records", 2, file.file().recordCount());
+    assertThat(file.file().path())
+        .as("Data file should be the table's manifest")
+        .isEqualTo(
+            Iterables.getOnlyElement(table.currentSnapshot().allManifests(table.io())).path());
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestEntriesMetadataTable.java:74]: can we express this check without using `Iterables`?

File: core/src/test/java/org/apache/iceberg/TestFindFiles.java [hunk 6]
```diff
@@ -18,40 +18,37 @@
...
-
-  public TestFindFiles(int formatVersion) {
-    super(formatVersion);
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(ParameterizedTestExtension.class)
+public class TestFindFiles extends TestBase {
+  @Parameters(name = "formatVersion = {0}")
+  public static List<Object> parameters() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFindFiles.java:38]:
```suggestion
  protected static List<Object> parameters() {
```

File: core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java [hunk 7]
```diff
@@ -31,49 +35,42 @@
...
-  public TestIncrementalDataTableScan(int formatVersion) {
-    super(formatVersion);
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(ParameterizedTestExtension.class)
+public class TestIncrementalDataTableScan extends TestBase {
+  @Parameters(name = "formatVersion = {0}")
+  public static List<Object> parameters() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java:45]:
```suggestion
  protected static List<Object> parameters() {
```
this was most likely previously required to be public by JUnit4, but that's not the case anymore with JUnit5, so we can make all of the parameters() methods protected

File: core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java [hunk 8]
```diff
@@ -196,56 +193,66 @@ public void testRollbacks() {
...
+  @TestTemplate
   public void testIgnoreResiduals() throws IOException {
     add(table.newAppend(), files("A"));
     add(table.newAppend(), files("B"));
     add(table.newAppend(), files("C"));
 
-    TableScan scan1 = table.newScan().filter(Expressions.equal("id", 5)).appendsBetween(1, 3);
+    IncrementalAppendScan scan1 =
+        table
+            .newIncrementalAppendScan()
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java:209]: why has this been changed?

File: core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java [hunk 9]
```diff
@@ -196,56 +193,66 @@ public void testRollbacks() {
...
     add(table.newAppend(), files("C"));
 
     AtomicInteger planThreadsIndex = new AtomicInteger(0);
-    TableScan scan =
+    IncrementalAppendScan scan =
         table
-            .newScan()
-            .appendsAfter(1)
+            .newIncrementalAppendScan()
+            .fromSnapshotExclusive(1)
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java:255]: why was this changed?

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 10]
```diff
@@ -356,14 +353,14 @@ public void testPartitionsTableScanNoFilter() {
...
             .asStruct();
     TableScan scanNoFilter = partitionsTable.newScan().select("partition.data_bucket");
-    Assert.assertEquals(expected, scanNoFilter.schema().asStruct());
+    assertThat(scanNoFilter.schema().asStruct()).isEqualTo(expected);
 
     CloseableIterable<ManifestEntry<?>> entries =
         PartitionsTable.planEntries((StaticTableScan) scanNoFilter);
     if (formatVersion == 2) {
-      Assert.assertEquals(8, Iterators.size(entries.iterator()));
+      assertThat(entries.iterator()).toIterable().hasSize(8);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:361]: can't we just use `assertThat(entries).hasSize(xx)` here?

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 11]
```diff
@@ -382,13 +379,13 @@ public void testPartitionsTableScanWithProjection() {
             .asStruct();
 
     TableScan scanWithProjection = partitionsTable.newScan().select("file_count");
-    Assert.assertEquals(expected, scanWithProjection.schema().asStruct());
+    assertThat(scanWithProjection.schema().asStruct()).isEqualTo(expected);
     CloseableIterable<ManifestEntry<?>> entries =
         PartitionsTable.planEntries((StaticTableScan) scanWithProjection);
     if (formatVersion == 2) {
-      Assert.assertEquals(8, Iterators.size(entries.iterator()));
+      assertThat(entries.iterator()).toIterable().hasSize(8);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:386]: same as above

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 12]
```diff
@@ -1357,27 +1344,27 @@ private void testPositionDeletesBaseTableFilter(boolean transactional) {
...
-        "Expected correct delete file on constant column",
-        FILE_A_DELETES.path(),
-        constantsMap(posDeleteTask, partitionType).get(MetadataColumns.FILE_PATH.fieldId()));
+    assertThat(filePartition).as("Expected correct partition on task's file").isEqualTo(0);
+    assertThat(taskPartition).as("Expected correct partition on task's column").isEqualTo(0);
+
+    assertThat(posDeleteTask.file().specId())
+        .as("Expected correct partition spec id on task")
+        .isEqualTo(0);
+    assertThat(constantsMap(posDeleteTask, partitionType).get(MetadataColumns.SPEC_ID.fieldId()))
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:1353]: this should be using `.containsEntry()`. Same for the other places in this PR

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScansWithPartitionEvolution.java [hunk 13]
```diff
@@ -193,38 +190,36 @@ public void testPositionDeletesPartitionSpecRemoval() {
...
+        .isEqualTo(1);
+
+    assertThat(posDeleteTask.spec().fields().get(0).fieldId())
+        .as("Expected correct partition field id on task's spec")
+        .isEqualTo(table.ops().current().spec().partitionType().fields().get(0).fieldId());
+
+    assertThat(posDeleteTask.file().specId())
+        .as("Expected correct partition spec id on task")
+        .isEqualTo(table.ops().current().spec().specId());
+    assertThat(constantsMap(posDeleteTask, partitionType).get(MetadataColumns.SPEC_ID.fieldId()))
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScansWithPartitionEvolution.java:211]: this should use .containsEntry()

**Comment by tomtongue** [comment]
Thanks for the review. Will reflect on the comments.

File: core/src/test/java/org/apache/iceberg/TestEntriesMetadataTable.java (hunk 5 above)
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestEntriesMetadataTable.java:74]: Sure, replace it with `get(0)`.

File: core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java (hunk 7 above)
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java:45]: Okay sure, thanks for the advice.

File: core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java (hunk 8 above)
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java:209]: `table.newScan(),appendsBetween` is marked as "deprecated" in https://iceberg.apache.org/javadoc/latest/org/apache/iceberg/TableScan.html so that I changed here.
> Deprecated.
since 1.0.0, will be removed in 2.0.0; use [Table.newIncrementalAppendScan()](https://iceberg.apache.org/javadoc/latest/org/apache/iceberg/Table.html#newIncrementalAppendScan()) instead.

But maybe this should be another PR, or keep it this time. Let me revert it to the previous version.

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java (hunk 10 above)
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:361]: This and other parts can be `assertThat(entries).hasSize(x)`. Will replace them.

File: core/src/test/java/org/apache/iceberg/TestBatchScans.java (hunk 3 above)
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestBatchScans.java:56]: Got it, thank you. I will create another PR for that migration. For now, keep the original one.

File: core/src/test/java/org/apache/iceberg/TestBase.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestBase.java:749]: Sure, thanks. Remove this method.

File: core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java [hunk 14]
```diff
@@ -196,56 +193,66 @@ public void testRollbacks() {
...
+              .isNotEqualTo(Expressions.alwaysTrue());
         }
       }
     }
 
-    TableScan scan2 =
-        table.newScan().filter(Expressions.equal("id", 5)).appendsBetween(1, 3).ignoreResiduals();
+    IncrementalAppendScan scan2 =
+        table
+            .newIncrementalAppendScan()
```
[the remark this answers is not in the document]
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java:227]: As I commented above, `table.newScan().appendsBetween` is deprecated. Revert this part to the previous one.

File: core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java (hunk 9 above)
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestIncrementalDataTableScan.java:255]: As I commented above, `table.newScan().appendsBetween` is deprecated. Revert this part to the previous one.

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java (hunk 12 above)
Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java]: Update the map tests with `.containsEntry()`. However each value is specified with the wildcard type like `capture of ?`, so the `contantsMap` value type needs to be changed in each test.

Comment (tomtongue) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java]: If there's a better idea, please let me know.

**Comment by tomtongue** [comment]
@nastra Update with your comments. When you have time, could you review the new commit?

File: core/src/test/java/org/apache/iceberg/TestFindFiles.java [hunk 15]
```diff
@@ -184,33 +181,33 @@ public void testCaseSensitivity() {
...
+  @TestTemplate
   public void testNoSnapshot() {
     // a table has no snapshot when it just gets created and no data is loaded yet
 
     // if not handled properly, NPE will be thrown in collect()
     Iterable<DataFile> files = FindFiles.in(table).collect();
 
     // verify an empty collection of data file is returned
-    Assert.assertEquals(0, Sets.newHashSet(files).size());
+    assertThat(Sets.newHashSet(files)).hasSize(0);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFindFiles.java:210]:
```suggestion
    assertThat(files).hasSize(0);
```

**Review by nastra — APPROVED** [review]
LGTM with one minor comment

**Comment by tomtongue** [comment]
@nastra Thank you! Reflected the minor comment. If there's any part that I should change, please let me know.
